src/System.Net.Security/src/System/Net/_SecureChannel.cs

```diff
@@ -0,0 +1,1487 @@
+                    }
+                }
+                finally
+                {
+                    if (issuerList.aIssuers != null)
+                    {
+                        issuerList.aIssuers.Dispose();
+                    }
+                }
+            }
+            return issuers;
+        }
+
+        /*++
+            AcquireCredentials - Attempts to find Client Credential
+            Information, that can be sent to the server.  In our case,
+            this is only Client Certificates, that we have Credential Info.
+
+            How it works:
+                case 0: Cert Selection delegate is present
+                        Always use its result as the client cert answer.
+                        Try to use cached credential handle whenever feasible.
+                        Do not use cached anonymous creds if the delegate has returned null
+                        and the collection is not empty (allow responding with the cert later).
+
+                case 1: Certs collection is empty
+                        Always use the same statically acquired anonymous SSL Credential
+
+                case 2: Before our Connection with the Server
+                        If we have a cached credential handle keyed by first X509Certificate
+                        **content** in the passed collection, then we use that cached
+                        credential and hoping to restart a session.
+
+                        Otherwise create a new anonymous (allow responding with the cert later).
+
+                case 3: After our Connection with the Server (i.e. during handshake or re-handshake)
+                        The server has requested that we send it a Certificate then
+                        we Enumerate a list of server sent Issuers trying to match against
+                        our list of Certificates, the first match is sent to the server.
+
+                        Once we got a cert we again try to match cached credential handle if possible.
+                        This will not restart a session but helps minimizing the number of handles we create.
+
+                In the case of an error getting a Certificate or checking its private Key we fall back
+                to the behavior of having no certs, case 1.
+
+            Returns: True if cached creds were used, false otherwise.
+
+        --*/
+
+        private bool AcquireClientCredentials(ref byte[] thumbPrint)
+        {
+            GlobalLog.Enter("SecureChannel#" + Logging.HashString(this) + "::AcquireClientCredentials");
+
+            // Acquire possible Client Certificate information and set it on the handle.
+            X509Certificate clientCertificate = null;   // This is a candidate that can come from the user callback or be guessed when targeting a session restart.
+            ArrayList filteredCerts = new ArrayList();  // This is an intermediate client certs collection that try to use if no selectedCert is available yet.
```

Comment:
Agreed but optimizations are out of scope for this initial merge. Tracking by #3114